getkirby/kirby
# New LockException and Lock Dialog

## Description

### Merge first

- https://github.com/getkirby/kirby/pull/6833

### Summary of changes

- New `Kirby\Content\LockException`
- New `k-lock-dialog`
- New `panel.content.lockDialog` method

### Reasoning

When an editor starts editing, but someone else made changes at the same time, there can be a race condition where the editor can no longer save or publish the changes because the other one was faster. This is now handled with a new dialog to inform the editor in a good way and reload the view afterwards. 

The new LockException is used to throw a custom error with all the lock details. This is used in the Panel to fill the dialog with the necessary information.

![Screenshot 2024-12-05 at 16 10 36](https://github.com/user-attachments/assets/7931c137-dfe6-48c4-8d8f-c43aa9f8e24a)

### Testing

The easiest way to test this is to make sure that you have two users. Let's assume, the second one has the id "editor"

1. Go to any page and start editing
2. Navigate to the changes text file in the filesystem and change the Lock: field from your user id to the "editor" user id.
3. Reload the page in the panel. The page should now be locked
4. Open the browser console and try to still save or publish changes with the following commands: 
```js
panel.content.save({subheading: 'Test'});
panel.content.publish({subheading: 'Test'});
```
The new dialog should appear and give you the right information about the lock. Closing the dialog should reload the view.
